### PR TITLE
Adding MSAL Client

### PR DESCRIPTION
**Pull request by lucidvoci**
#### What type of PR is this?

Addition of MSAL library in order to be able authenticate using the dSTS token.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+				[]string{"https://management.azure.com/.default"},
+				confidential.WithTenantID(ba.tenantID),
+			)
+
+			if errT != nil {
+				return nil, fmt.Errorf("failed to retrieve a token: %w", errT)
+			}
+
+			r, err := p.Prepare(r)
+			if err == nil {
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: this should be used only exceptionally - e.g. if you have a method which entire body would be the err path and then one line of the non-err path, it's preferable to do `err == nil` at the top and then handle the err path, but in here I just got confused thinking that we are doing `.Prepare` in case of an error

example of a code where I would do `err == nil` over `err != nil`

```go
func foo() {
  bar, err := something()
  if err != nil {
    super not interesting many lines of error handling
    something
    stuff
    lorem ipsum
    extremely long
    if more indent {
      bar
    }
    20 lines of code more..
  }
  return true
```

this would read better with `err == nil`

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+				return autorest.Prepare(r, autorest.WithHeader("Authorization", fmt.Sprintf("Bearer %s", token.AccessToken)))
+			}
+			return r, err
+		})
+	}
+}
+
+func NewMSALClient(config *Config) (confidential.Client, error) {
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:35]: nit: 'idiomatic' way of structuring go files usually is

```
struct
constructor for struct
exported struct methods
unexported struct methods
exported free functions
unexported free functions
```

you're going to hear me saying 'idiomatic' a lot, sorry, golang is like that

tl;dr: I would put this above the `WithAuthorization` method

Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:35]: I just checked the code for `newServicePrincipalTokenFromCredentials` and there are bunch of code-paths that let the caller create credentials from

```
if config.UseWorkloadIdentityExtension
if config.UseManagedIdentityExtension
if len(config.AADClientSecret) > 0 {
if len(config.AADClientCertPath) > 0 && len(config.AADClientCertPassword) > 0 {
```

can we initialize MSAL flow via these as well? if yes, we need to add those, otherwise this won't get accepted

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return confidential.Client{}, fmt.Errorf("couldn't read cert file %s: %w", config.AADClientCertPath, err)
+	}
+
+	cert, key, err := azidentity.ParseCertificates(b, nil)
+	if err != nil {
+		return confidential.Client{}, fmt.Errorf("couldn't parse certificate: %w", err)
+	}
+
+	authorityUrl, err := url.JoinPath(config.Location, config.TenantID)
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:46]: nit: this is used on line 90, I would move it closer to there, preferably above

rule of thumb would be: initialize variables as close to their usage as possible

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	cred, err := confidential.NewCredFromCert(cert, key)
+	if err != nil {
+		return confidential.Client{}, fmt.Errorf("failed to create creds object from certificate+key: %w", err)
+	}
+
+	opts := []confidential.Option{
+		// confidential.WithAzureRegion(confidential.AutoDetectRegion()),
+		// confidential.WithKnownAuthorityHosts([]string{authorityHost}), // tmp
+		// confidential.WithInstanceDiscovery(false),                     // tmp
+	}
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: unused code, remove pls

File: cluster-autoscaler/go.mod [hunk 5]
```diff
@@ -1,24 +1,28 @@
 module k8s.io/autoscaler/cluster-autoscaler
 
-go 1.21
+go 1.21.3
+
+toolchain go1.21.4
```
Comment (handsomejack-42) [inline cluster-autoscaler/go.mod]: I don't think we nede to update the go version in this PR, same for adding `toolchain` directive

please roll back

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// WithAuthorization returns a PrepareDecorator that adds an HTTP Authorization header whose
+// value is "Bearer " followed by the token.
+func (ba *MSALBearerAuthorizer) WithAuthorization() autorest.PrepareDecorator {
+	return func(p autorest.Preparer) autorest.Preparer {
+		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
+			// TODO: get refresh token when available
+			token, errT := ba.client.AcquireTokenByCredential(
+				context.Background(),
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: please use context from incoming request (`r.Context()`)

if you need explaining what a context in go is and what is it for, let's discuss f2f, it's a beast and I'm too lazy to write 5 paragraphs here :see_no_evil:

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+				[]string{"https://management.azure.com/.default"},
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: my understanding of MSAL is still vague, but I do think this can't be static - that only works for our dSTS PoC, the scopes need to be set by a variable ~somehow

most likely in `MSALBearerAuthorizer` using `With*` options - let's sync on that

I believe the way forward would be to go and check e.g. azure-sdk-for-go, which already uses MSAL pretty extensively and see what capabilities in terms of configuration it provides.. I am concerned that we might have to support the same init capabilities here

File: cluster-autoscaler/cloudprovider/azure/azure_client_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMSALAuthorizer(t *testing.T) {
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_client_test.go]: in the final version of this PR, this should be either removed or replaced with a proper, much more complicated test - atm, this is just a way for you to be able to run it, it's not a test per-se

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	tenantID string
+}
+
+// WithAuthorization returns a PrepareDecorator that adds an HTTP Authorization header whose
+// value is "Bearer " followed by the token.
+func (ba *MSALBearerAuthorizer) WithAuthorization() autorest.PrepareDecorator {
+	return func(p autorest.Preparer) autorest.Preparer {
+		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
+			// TODO: get refresh token when available
+			token, errT := ba.client.AcquireTokenByCredential(
```
Comment (handsomejack-42) [inline cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: why `errT` ? please use `err` unless it collides and you need to avoid shadowing (which is usually fine, just shadow, it's `err`)

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go (hunk 3 above)
Comment (lucidvoci) [reply cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:46]: The cred and opts are also initialized just for the confidence.New method, so doesn't it kinda follows this principle already?

Comment (handsomejack-42) [reply cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:46]: I didn't notice that `cred` exists for the same purpose, sorry

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go (hunk 7 above)
Comment (lucidvoci) [reply cluster-autoscaler/cloudprovider/azure/azure_msal_client.go]: This issue was added to [PIB 2739552](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2739552) and will be revisited when the implementation of MSAL is continued.

File: cluster-autoscaler/cloudprovider/azure/azure_msal_client.go (hunk 2 above)
Comment (lucidvoci) [reply cluster-autoscaler/cloudprovider/azure/azure_msal_client.go:35]: This issue was added to [PIB 2739552](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2739552) and will be revisited when the implementation of MSAL is continued.
